### PR TITLE
[feature] STLink-v3: Add stlink v3 support

### PR DESCRIPTION
**Pull request by Nightwalker-87**

### CODE REVIEW

**Comment by Nightwalker-87** [comment]
Merged into develop branch with commit aad5cf1901f467914a2efe855c0caff7fdf99048.
Related to issue #820.

**Comment by bmarvo** [comment]
@Nightwalker-87 did you merge this into develop manually? What is the status of this PR? There are conflicts.

**Comment by Nightwalker-87** [comment]
@bmarvo: Yes, and I'm aware of that. I'll take a closer look at it after my next outstanding commit.

**Comment by Nightwalker-87** [comment]
Conflicts resolved and content merged with PR #866. Closes #863.

**Comment by Nightwalker-87** [comment]
@martonmiklos: Thx for your contribution! :+1:

File: src/usb.c [hunk 1]
```diff
@@ -334,17 +394,23 @@ int _stlink_usb_enter_swd_mode(stlink_t * sl) {
...
     unsigned char* const cmd  = sl->c_buf;
     ssize_t size;
-    const int rep_len = 0;
+    unsigned char* const data = sl->q_buf;
+    const uint32_t rep_len = sl->version.stlink_v > 1 ? 2 : 0;
     int i = fill_command(sl, SG_DXFER_FROM_DEV, rep_len);
 
     cmd[i++] = STLINK_DEBUG_COMMAND;
-    cmd[i++] = STLINK_DEBUG_ENTER;
+    cmd[i++] = sl->version.stlink_v > 1? STLINK_DEBUG_APIV2_ENTER : STLINK_DEBUG_ENTER;
```
Comment (Nightwalker-87) [inline src/usb.c:402]: @martonmiklos: This fixes issue 271 doesn't it?

**Comment by Nightwalker-87** [comment]
@martonmiklos: It seems I have messed up proper merging of this last time.
Can you resolve the conflicts in your branch to allow for a clean merge?
You may also add further commits, should you have made any progress on this topic since then.

**Comment by martonmiklos** [comment]
Well for me it was quite a surprise that you merged the V3 support upstream, I just stated in the #820 that some testing needed not that it is finished. Anyway do not worry, I will figure this out.

**Comment by grevaillot** [comment]
i've been using a stlinkv3 recently, on a g474 board, and except a strange trace at start :

> INFO common.c: *** stlink_force_debug_mode ***
> INFO common.c: *** stlink_status ***
>  core status: unknown

that need to be taken care of, i've had no catastrophic issues.

File: src/usb.c (hunk 1 above)
Comment (Nightwalker-87) [inline src/usb.c:402]: This fixes #271.

**Comment by Nightwalker-87** [comment]
@martonmiklos: Could you add a missing comment related to #271 in `/src/usb.c`? 
Any further news on this topic?

**Comment by Nightwalker-87** [comment]
I've just found that merge-commit aad5cf1901f467914a2efe855c0caff7fdf99048 accidentally introduced a bug on my STlink-v2 (clone) device. Entering `st-info --probe` gives the following output:

```
[!] send_recv read reply failed: LIBUSB_ERROR_PIPE
[!] send_recv STLINK_DEBUG_RESETSYS
[!] send_recv read reply failed: LIBUSB_ERROR_PIPE
[!] send_recv STLINK_DEBUG_READCOREID
[!] send_recv read reply failed: LIBUSB_ERROR_PIPE
[!] send_recv STLINK_JTAG_READDEBUG_32BIT
Found 1 stlink programmers
```
in contrast to the previous commit:

```
Found 1 stlink programmers
```

This error is reproducible as often as is retry, recompile or plug/unplug the device, so one can assume something got screwed up there. I haven't found the exact place in the code yet, but fortunately we already know exactly in which commit it appeared...

**Comment by martonmiklos** [comment]
@Nightwalker-87 

Well on Linux with a clone it works correctly:

```
mm@lapos:/$  cd /home/mm/Projektek/stlink
mm@lapos:~/Projektek/stlink$ git branch 
* add_stlink_v3_support
  master
mm@lapos:~/Projektek/stlink$ git describe                                                                                                                                              
v1.6.0-6-ge81f23e 
mm@lapos:~/Projektek/stlink$ git diff
mm@lapos:~/Projektek/stlink$ cd build/
mm@lapos:~/Projektek/stlink/build$ cmake ..
-- stlink version: 1.6.0-6-ge81f23e
--           Major 1 Minor 6 Patch 0
-- found USB
-- Manpage generation disabled
-- Debian-based Linux OS detected
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mm/Projektek/stlink/build
mm@lapos:~/Projektek/stlink/build$ make
[ 20%] Built target stlink-static
[ 41%] Built target stlink
[ 47%] Built target st-info
[ 55%] Built target st-flash
[ 67%] Built target st-util
[ 73%] Built target stlink-gui
[ 79%] Built target stlink-gui-local
[ 85%] Built target sg
[ 94%] Built target flash
[100%] Built target usb
mm@lapos:~/Projektek/stlink/build$ ./st-info --probe
Found 1 stlink programmers
mm@lapos:~/Projektek/stlink/build$ lsusb 
...
Bus 002 Device 010: ID 0483:3748 STMicroelectronics ST-LINK/V2
....
```

**Comment by Nightwalker-87** [comment]
I think the seen error derived from the manual merge into the master branch, while your personal branch remained clean. So your findings don't really surprise me. You would need to check out the commit I mentioned earlier.

File: include/stlink/usb.h [hunk 2]
```diff
@@ -18,11 +18,15 @@
 extern "C" {
 #endif
 
-#define STLINK_USB_VID_ST               0x0483
-#define STLINK_USB_PID_STLINK           0x3744
-#define STLINK_USB_PID_STLINK_32L       0x3748
-#define STLINK_USB_PID_STLINK_32L_AUDIO 0x374a
-#define STLINK_USB_PID_STLINK_NUCLEO    0x374b
+#define STLINK_USB_VID_ST                   0x0483
```
Comment (Nightwalker-87) [inline include/stlink/usb.h:21]: Indentation?

**Comment by Nightwalker-87** [comment]
@martonmiklos: Thank you. This is a lot better now. :+1: 
However there are still some compile errors with clang and on macOS that want to be addressed.

**Comment by chenguokai** [comment]
The current compile errors are again associated with the 64 to 32 conversion. May need a unified way of dealing with size_t.

**Comment by Nightwalker-87** [comment]
Can we turn this PR into a working state to have it merged?

**Comment by Nightwalker-87** [comment]
Can we turn this PR into a working state to have the present changes merged?
If no further interest appears (resulting in somehow active contribution) to fully implement this feature for the upcoming release, then we will leave it at that and push the related ticket to a later milestone. This feature will _not_ block our next release, nor should this PR remain open for much longer. Also orphaned PRs for incomplete features do not comply with our contribution guidelines.

**Comment by Nightwalker-87** [comment]
Merged into `stlink-v3_pre` branch with commit 442ade88ddce1fc8ce4d36c856b57c24ee76aa12.

**Comment by Nightwalker-87** [comment]
> The current compile errors are again associated with the 64 to 32 conversion. May need a unified way of dealing with size_t. Too many casts are really a burden.

@chenguokai: Indeed. I'm currently trying to deal with this. Maybe a change in `stlink.h` could be the key to success. Nevertheless suggestions are welcome.

@martonmiklos: We also need to know what what has been addressed here (short summary or listing) and what parts are still missing as of the current state of knowledge. You should know best.

**Comment by martonmiklos** [comment]
ST implemented two different JTAG command API version: V1 and V2. See here APIV1 suffixed commands in the OpenOCD code:
https://github.com/ntfreak/openocd/blob/master/src/jtag/drivers/stlink_usb.c#L227

The stlink uses only the V1 commands and it seems that ST dropped the support for them in the STLinkV3 firmware.

As it turns out from the openOCD code the V2 API was even supported by the STLinkV1 above a certain version:
https://github.com/ntfreak/openocd/blob/master/src/jtag/drivers/stlink_usb.c#L991

Apart from utilising the V2 command API there are some other minor differences in the V3 handling like extended version query, which generally seems to do it's job. 

At the moment most of the commands are working, however the flash loaded never hits a breakpoint after loading as I mentioned somewhere else (sometimes for me it is hard to follow the multiple PRs/issues associated with the V3).

I have traced the SWD with a logic analyzer where I see a small amount of difference between the V2/V3 traces. However I would need to route the stlink cmdline output to a serial line which could be used to track down where exactly happens the difference.

**Comment by Nightwalker-87** [comment]
>  I have traced the SWD with a logic analyzer where I see a small amount of difference between the V2/V3 traces. However I would need to route the stlink cmdline output to a serial line which could be used to track down where exactly happens the difference.

We an address this within #820 at a later point in time. From your preceding explanations I read that basic support/compatibility seems to be present already (though not sufficiently tested yet). I have added the key points of your descriptions to a description for an upcoming PR which covers all changes that have been made here and will include a solution for the 64 to 32 conversion issue to ensure all build tests will pass.

**Comment by chenguokai** [comment]
> Maybe a change in stlink.h could be the key to success. Nevertheless suggestions are welcome.

I would suggest unifying the return value types. For example, a `ssize_t` return value conflicts with an `int` variable. In most cases, an `int` works well enough, since `ssize_t` itself is variable in length and contains sign info, just like a simple `int`. When a fixed length variable is in need, `uintX_t` or `intX_t` will do.

**Comment by Nightwalker-87** [comment]
I left a note on that in #954.
Thus we should close this long conversation here now, as all points have now been addressed.
